garden_ai/modal/functions.py

```diff
@@ -1,61 +1,113 @@
-import asyncio
-from concurrent.futures import ThreadPoolExecutor
-from typing import TYPE_CHECKING, Any, TypeVar
+import io
+from typing import TYPE_CHECKING, TypeVar
 
-import modal
 from modal._serialization import serialize
-from modal._utils.blob_utils import MAX_OBJECT_SIZE_BYTES
+from modal._utils.async_utils import synchronize_api
+from modal._utils.blob_utils import (
+    DEFAULT_SEGMENT_CHUNK_SIZE,
+    MAX_OBJECT_SIZE_BYTES,
+    BytesIOSegmentPayload,
+    _download_from_url,
+    _upload_to_s3_url,
+    get_content_length,
+    perform_multipart_upload,
+)
 from modal._utils.function_utils import _process_result
-from modal_proto import api_grpc, api_pb2  # type: ignore
+from modal._utils.hash_utils import get_upload_hashes
+from modal_proto import api_pb2  # type: ignore
 
 if TYPE_CHECKING:
     from garden_ai.client import GardenClient
 else:
     GardenClient = TypeVar("GardenClient")
 
 from ..schemas.modal import (
+    ModalBlobUploadURLRequest,
     ModalFunctionMetadata,
     ModalInvocationRequest,
     ModalInvocationResponse,
 )
 
 
-class MaximumArgumentSizeError(ValueError):
-    """Raised when input arguments are too large to pass directly to modal"""
-
-    pass
-
-
-class ModalFunction:
+class _ModalFunction:
     def __init__(
         self, metadata: ModalFunctionMetadata, client: GardenClient | None = None
     ):
-        self.metadata = metadata
-        self.client = client
-        if self.client is None:
-            from garden_ai import GardenClient
-
-            self.client = GardenClient()
+        self._metadata = metadata
+        self._client = client
+
+    @property
+    def metadata(self) -> ModalFunctionMetadata:
+        return self._metadata
+
+    @property
```

Comment:
reviewer note: the read-only `@property` style approach plays better with the synchronicity-wrapped class than plain init attributes did, but isn't doing anything special otherwise